File: transport/internal/auth/handshake.go
package auth

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"errors"
	"io"
	"net"
	"time"
)

func SecureHandshakeClient(conn net.Conn, authKey []byte) error {
	nonce := make([]byte, 32)
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return err
	}

	mac := hmac.New(sha256.New, authKey)
	mac.Write(nonce)
	signature := mac.Sum(nil)

	payload := append(nonce, signature...)

	conn.SetWriteDeadline(time.Now().Add(5 * time.Second))
	if _, err := conn.Write(payload); err != nil {
		return err
	}
	conn.SetWriteDeadline(time.Time{})

	return nil
}

func SecureHandshakeServer(conn net.Conn, authKey []byte) error {
	payload := make([]byte, 64)

	conn.SetReadDeadline(time.Now().Add(5 * time.Second))
	if _, err := io.ReadFull(conn, payload); err != nil {
		return err
	}
	conn.SetReadDeadline(time.Time{})

	nonce := payload[:32]
	signature := payload[32:]

	mac := hmac.New(sha256.New, authKey)
	mac.Write(nonce)
	expectedSignature := mac.Sum(nil)

	if !hmac.Equal(signature, expectedSignature) {
		return errors.New("invalid signature")
	}

	return nil
}

File: libtransport/src/handshake.rs
use std::io::{self, Read, Write};
use std::net::TcpStream;
use std::time::Duration;

use hmac::{Hmac, Mac};
use sha2::Sha256;
use zeroize::Zeroize;

type HmacSha256 = Hmac<Sha256>;

const HANDSHAKE_TIMEOUT: Duration = Duration::from_secs(5);

pub fn client(stream: &mut TcpStream, auth_key: &[u8; 32]) -> io::Result<()> {
    stream.set_write_timeout(Some(HANDSHAKE_TIMEOUT))?;

    let mut nonce = [0u8; 32];
    getrandom::getrandom(&mut nonce).map_err(|_| io::Error::new(io::ErrorKind::Other, "rng failure"))?;

    let mut mac = HmacSha256::new_from_slice(auth_key)
        .map_err(|e| io::Error::new(io::ErrorKind::InvalidInput, e))?;
    mac.update(&nonce);
    let signature = mac.finalize().into_bytes();

    let mut payload = [0u8; 64];
    payload[..32].copy_from_slice(&nonce);
    payload[32..].copy_from_slice(&signature);

    stream.write_all(&payload)?;
    stream.flush()?;

    nonce.zeroize();

    stream.set_write_timeout(None)?;
    Ok(())
}

pub fn server(stream: &mut TcpStream, auth_key: &[u8; 32]) -> io::Result<()> {
    stream.set_read_timeout(Some(HANDSHAKE_TIMEOUT))?;

    let mut payload = [0u8; 64];
    stream.read_exact(&mut payload)?;

    let nonce = &payload[..32];
    let signature = &payload[32..];

    let mut mac = HmacSha256::new_from_slice(auth_key)
        .map_err(|e| io::Error::new(io::ErrorKind::InvalidInput, e))?;
    mac.update(nonce);

    mac.verify_slice(signature)
        .map_err(|_| io::Error::new(io::ErrorKind::PermissionDenied, "invalid auth"))?;

    payload.zeroize();
    stream.set_read_timeout(None)?;
    Ok(())
}
